Refactor matocsserv files

ralcolea opened the pull request:
This PR refactors matocsserv files to:
- Remove custom C-linked lists used across these files to handle replication and chunkserver connections with the master (matocsserventry).
- Remove functions related with the allocation of memory and traversing of previous C-linked lists, making the code cleaner and easier to understand and maintain.
- Remove redundant `void`, `return` and `else` instructions.
- Remove some commented logs that remained in the code (dead code).
- Replace `std::vector::push_back()` functions by `std::vector::emplace_back()` where it was possible.

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
refactor(master): Refactor matocsserventry list

This commit replaces C-linked list of raw matocsserventry* pointers
with a std::list of smart pointers. This simplifies memory management
by ensuring automatic deallocation and improves maintainability for
operations such as traversal and modification.

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
refactor(master): Modernize replication lists

This commit modernizes the in-memory structures used for tracking
replication state in matocsserv.cc. It replaces custom C-style linked
lists and manual memory management (malloc, free) with std::list and
dynamic allocation via new.

The key changes are:
- Simplifies memory ownership and lifecycle management.
- Removes custom C-style linked lists for repsrc and repdst structures.
- Replaces replicas hash bucket of raw pointers (rephash) by hash
  bucket of std::list<repdst*>.
- Improves readability and safety during iteration and modification of
  replication state.

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
refactor(master): Minor refactor of matocsserv.cc

This commit applies a minor refactor to address some warnings and
redundant instructions.

Key relevant changes:
- Apply format style in the signature of the modified functions.
- Rename some parameters to use more meaningful names.
- Replace std::vector::push_back() functions by emplace_back() where
  it's possible.
- Removed redundant `void`, `return` and `else` instructions.
- Removed dead code, like commented logs.

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
chore: Remove unused function

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
fix(master): Fix memory leaks in matocsserv files

Signed-off-by: Crash <crash@leil.io>

ralcolea commented on src/master/matocsserv.cc:
```diff
@@ -220,120 +184,113 @@ int matocsserv_replication_find(uint64_t chunkId, uint32_t chunkVersion,
-			*replicaPointer = replica->next;
-			matocsserv_repdst_free(replica);
+
+			replica->replicaSourceList.clear();
```
Nice catch :)
This leak was solved as you suggested, using `std::list<std::unique_ptr<repsrc>>`.

ralcolea commented on src/master/matocsserv.cc:
```diff
@@ -220,120 +184,113 @@
+			replica->replicaSourceList.clear();
+			destination->wrepcounter--;
+
+			replicaIt = rephash[hash].erase(replicaIt);
```
Nice catch :)
This leak was solved deleting the object before erasing the pointer from the list.

ralcolea commented on src/master/matocsserv.cc:
```diff
@@ -220,120 +184,113 @@
-				*rp = r->next;
-				matocsserv_repdst_free(r);
+
+				replica->replicaSourceList.clear();
```
Nice catch :)
This leak was solved as you suggested, using std::list<std::unique_ptr<repsrc>>.

ralcolea commented on src/master/matocsserv.cc:
```diff
@@ -220,120 +184,113 @@
+				replica->replicaSourceList.clear();
+				server->wrepcounter--;
+
+				replicaIterator = hash.erase(replicaIterator);
```
Nice catch :)
This leak was solved deleting the object before erasing the pointer from the list.

ralcolea commented on src/master/matocsserv.cc:
```diff
@@ -220,120 +184,113 @@
+					auto *replicaSource = *replicaSrcIterator;
+					if (replicaSource->src == server) {
+						server->rrepcounter--;
+						replicaSrcIterator = replica->replicaSourceList.erase(replicaSrcIterator);
```
Nice catch :)
This leak was solved as you suggested, using std::list<std::unique_ptr<repsrc>>.

dmga44 commented on src/master/matocsserv.cc:
```diff
@@ -314,26 +318,27 @@ void matocsserv_replication_disconnected(matocsserventry *srv) {
-				} else if (space<minspace) {
-					minspace = space;
-				} else if (space>maxspace) {
-					maxspace = space;
+	for (const auto &eptr : matocsservList) {
+		if (j < 65535 && k < 65535) {
+			if (eptr->mode != KILL) {
+				if (eptr->totalspace > 0 && eptr->usedspace <= eptr->totalspace) {
+					space = (double)(eptr->usedspace) / (double)(eptr->totalspace);
+					if (j == 0) {
+						minspace = maxspace = space;
+					} else if (space < minspace) {
+						minspace = space;
+					} else if (space > maxspace) {
+						maxspace = space;
+					}
+					j++;
 				}
-				j++;
+				k++;
```
This is a minor suggestion. Could you consider refactoring function as well? For instance:
- j is the number of "consistent" chunkservers which conection is not supposed to be ended
- k is the number of chunkservers which connections are not supposed to be ended
- please notice that `j <= k` so the check could be simplified.
- the bunch of if/else if statements to calculate the minspace and maxspace could be simplified using std::min and std::max

dmga44 replied:
Also notice that previous implementation was stopping when reaching the `65535` alive chunkservers (probably due to overflow of the return types of those metrics) and the new implementation finishes the cycle around the list (might waste some time in those cases, but seems very unlikely case).

ralcolea replied:
I liked your suggestions @dmga44 :+1: 
I applied all your suggestions. Please, let me know your ideas about the latest changes.

dmga44 reviewed (approved):
Great job @ralcolea 👍

ralcolea committed:
chore: Apply code review suggestions

This commit addresses the suggestions requested by the reviewers.
The key changes are:
- Refactor `matocsserv_usagedifference()` function.
- Addition of smart pointers to the replicas hash bucket.

Co-authored-by: Dave <dave@leil.io>
Co-authored-by: guillex <guillex@leil.io>

Signed-off-by: Crash <crash@leil.io>
